File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    vector<int> nse(vector<int> arr) {
        int n = arr.size();
        vector<int> ans(n);
        stack<int> st;

        // Last element ka NSE hamesha n hota hai (right me koi nahi)
        ans[n - 1] = n;
        st.push(n - 1);

        for (int i = n - 2; i >= 0; i--) {
            while (!st.empty() && arr[st.top()] >= arr[i]) { // >= taaki equal heights bhi pop ho
                st.pop();
            }
            if (st.empty()) ans[i] = n;
            else ans[i] = st.top();
            st.push(i);
        }
        return ans;
    }

    vector<int> pse(vector<int> arr) { // previous smaller element
        int n = arr.size();
        vector<int> ans(n);
        stack<int> st;

        // First element ka PSE hamesha -1 hota hai (left me koi nahi)
        ans[0] = -1;
        st.push(0);

        for (int i = 1; i < n; i++) {
            while (!st.empty() && arr[st.top()] >= arr[i]) { // >= taaki equal heights bhi pop ho
                st.pop();
            }
            if (st.empty()) ans[i] = -1;
            else ans[i] = st.top();
            st.push(i);
        }
        return ans;
    }

    int largestRectangleArea(vector<int>& heights) {
        vector<int> nseidx = nse(heights);
        vector<int> pseidx = pse(heights);
        int n = nseidx.size();
        int ans = 0;

        for (int i = 0; i < n; i++) {
            int area = (nseidx[i] - 1 - pseidx[i]) * heights[i];
            ans = max(ans, area);
        }
        return ans;
    }
};

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights)
    {
        int left[]=previousSmaller(heights);
        int right[]=nextSmaller(heights);
        int maxArea=0;

        for(int i=0;i<heights.length;i++)
        {
            int width=right[i]-left[i]-1;
            int area=heights[i]*width;
            maxArea=Math.max(area,maxArea);
        }

        return maxArea;
    }

    private int[] nextSmaller(int nums[])
    {
        Stack<Integer> st=new Stack<>();
        int ans[]=new int[nums.length];
        for(int i=ans.length-1;i>=0;i--)
        {
            while(!st.isEmpty() && nums[st.peek()]>=nums[i])
            {
                st.pop();
            }
            if(st.isEmpty())
            {
                ans[i]=ans.length;
            }
            else 
            {
                ans[i]=st.peek();
            }

            st.push(i);
        }
        return ans;
    }

    private int[] previousSmaller(int nums[])
    {
            Stack<Integer> st=new Stack<>();
            int ans[]=new int[nums.length];
            for(int i=0;i<ans.length;i++)
            {
                while(!st.isEmpty() && nums[st.peek()]>=nums[i])
                {
                    st.pop();
                }
                if(st.isEmpty())
                {
                    ans[i]=-1;
                }
                else 
                {
                    ans[i]=st.peek();
                }

                st.push(i);
            }
            return ans;
    }
}